boa/src/syntax/parser/mod.rs

```diff
@@ -1354,100 +1447,106 @@ impl Parser {
 
     /// <https://tc39.es/ecma262/#prod-Arguments>
     fn read_arguments(&mut self) -> Result<Vec<Node>, ParseError> {
-        let mut args = vec![];
+        let mut args = Vec::new();
         loop {
-            match self.next_skip_lineterminator() {
-                Ok(ref tok) if tok.kind == TokenKind::Punctuator(Punctuator::CloseParen) => break,
-                Ok(ref tok) if tok.kind == TokenKind::Punctuator(Punctuator::Comma) => {
+            let next_token = self
+                .next_skip_lineterminator()
+                .ok_or(ParseError::AbruptEnd)?;
+            match next_token.kind {
+                TokenKind::Punctuator(Punctuator::CloseParen) => break,
+                TokenKind::Punctuator(Punctuator::Comma) => {
                     if args.is_empty() {
-                        return Err(ParseError::Unexpected(tok.clone(), None));
+                        return Err(ParseError::Unexpected(next_token.clone(), None));
                     }
-                    if self.next_if_skip_lineterminator(TokenKind::Punctuator(
-                        Punctuator::CloseParen,
-                    ))? {
+
+                    if self
+                        .next_if_skip_lineterminator(TokenKind::Punctuator(Punctuator::CloseParen))
+                        .is_some()
+                    {
                         break;
                     }
                 }
-                Ok(tok) => {
+                _ => {
                     if !args.is_empty() {
                         return Err(ParseError::Expected(
                             vec![
                                 TokenKind::Punctuator(Punctuator::Comma),
                                 TokenKind::Punctuator(Punctuator::CloseParen),
                             ],
-                            tok,
-                            "expected ',' or ')'.",
+                            next_token.clone(),
+                            Some("argument list"),
                         ));
                     } else {
-                        self.step_back();
+                        self.cursor.back();
                     }
                 }
-                Err(_) => {
-                    return Err(ParseError::AbruptEnd);
-                }
-            }
-            if self
-                .next_if(TokenKind::Punctuator(Punctuator::Spread))
-                .is_some()
-            {
-                args.push(Node::Spread(Box::new(self.read_assignment_expression()?)));
-            } else {
-                args.push(self.read_assignment_expression()?);
             }
+
+            args.push(self.read_assignment_expression()?);
```

Comment:
> It seems that this new implementation is giving a test error:
> 
> ```
> ---- exec::tests::spread_with_arguments stdout ----
> thread 'exec::tests::spread_with_arguments' panicked at 'assertion failed: `(left == right)`
>   left: `"undefined"`,
>  right: `"1"`', boa/src/exec/tests.rs:55:5
> ```
This will fix the failing test.
```suggestion

            if self
                .next_if(TokenKind::Punctuator(Punctuator::Spread))
                .is_some()
            {
                args.push(Node::Spread(Box::new(self.read_assignment_expression()?)));
            } else {
                args.push(self.read_assignment_expression()?);
            }
```